src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -751,46 +661,61 @@ public override Task WriteAsync(char[] buffer, int index, int count)
+                return base.WriteAsync(buffer, cancellationToken);
+            }
+
+            if (_stream == null)
+            {
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_WriterClosed);
+            }
+
+            CheckAsyncTaskInProgress();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            Task task = WriteAsyncInternal(this, buffer, _charBuffer, _charPos, _charLen, CoreNewLine, _autoFlush, appendNewLine: false, cancellationToken: cancellationToken);
+            _asyncWriteTask = task;
+            return task;
+        }
+
         // We pass in private instance fields of this MarshalByRefObject-derived type as local params
         // to ensure performant access inside the state machine that corresponds this async method.
         // Fields that are written to must be assigned at the end of the method *and* before instance invocations.
-        private static async Task WriteAsyncInternal(StreamWriter _this, char[] buffer, int index, int count,
+        private static async Task WriteAsyncInternal(StreamWriter _this, ReadOnlyMemory<char> source,
                                                      char[] charBuffer, int charPos, int charLen, char[] coreNewLine,
-                                                     bool autoFlush, bool appendNewLine)
+                                                     bool autoFlush, bool appendNewLine, CancellationToken cancellationToken)
         {
-            Debug.Assert(count == 0 || (count > 0 && buffer != null));
-            Debug.Assert(index >= 0);
-            Debug.Assert(count >= 0);
-            Debug.Assert(buffer == null || (buffer != null && buffer.Length - index >= count));
-
-            while (count > 0)
+            while (source.Length > 0)
             {
                 if (charPos == charLen)
                 {
-                    await _this.FlushAsyncInternal(false, false, charBuffer, charPos).ConfigureAwait(false);
+                    await _this.FlushAsyncInternal(false, false, charBuffer, charPos, cancellationToken).ConfigureAwait(false);
                     Debug.Assert(_this._charPos == 0);
                     charPos = 0;
                 }
 
-                int n = charLen - charPos;
-                if (n > count)
-                {
-                    n = count;
-                }
-
+                int n = Math.Min(charLen - charPos, source.Length);
                 Debug.Assert(n > 0, "StreamWriter::Write(char[], int, int) isn't making progress!  This is most likely a race condition in user code.");
 
-                Buffer.BlockCopy(buffer, index * sizeof(char), charBuffer, charPos * sizeof(char), n * sizeof(char));
+                bool copied = source.Span.Slice(0, n).TryCopyTo(new Span<char>(charBuffer, charPos, n));
+                Debug.Assert(copied);
```

Comment:
It looks like the only action when `copied` is false is to throw during debugging. Why not use CopyTo instead and throw on release as well? Otherwise, it seems like you do nothing with the source span if the user passes in a charPos != 0.

The only time TryCopyTo would fail is if charPos != 0, correct? Should that be the assertion before copying?

Reply:
> It looks like the only action when copied is false is to throw during debugging

To assert, yes.

> Why not use CopyTo instead and throw on release as well?

Because there's no situation in which this should happen, and using CopyTo incurs another check/branch/etc.  But it doesn't really matter; I can change to CopyTo if you prefer.